File: 19_mar/quickSort.java
class quickSort {
	public static void swap(int[] arr, int idx1, int idx2) {
		int temp = arr[idx1];
		arr[idx1] = arr[idx2];
		arr[idx2] = temp;
	}

	public static int partition(int[] arr, int left, int right) {
		int partIdx = left,
			partVal = arr[right];

		for (int i = left; i < right; i++) {
			if (arr[i] < partVal) {
				swap(arr, i, partIdx);
				partIdx++;
			}
		}

		swap(arr, partIdx, right);

		return partIdx;
	}

	public static void quickSort(int[] arr, int left, int right) {
		if (left < right) {
			int partIdx = partition(arr, left, right);

			quickSort(arr, left, partIdx - 1);
			quickSort(arr, partIdx + 1, right);
		}
	}

	public static void quickSort(int[] arr) {
		quickSort(arr, 0, arr.length - 1);
	}

	public static void main(String[] args) {
		int[] inpArr = {1,2,3,4,5,6,7,8,9};

		quickSort(inpArr);

		for (int i = 0; i < inpArr.length; i++) {
			System.out.print(" ");
			System.out.print(inpArr[i]);
		}
	}
}

File: js/dailies/05_aug_18/quicksort.js
const randomArr = require('../../helpers/randomArr');

const partition = (arr, begIdx, endIdx) => {
	const randIdx = Math.floor((endIdx - begIdx) * Math.random()) + begIdx;
	swap(arr, begIdx, randIdx);

	let partIdx = begIdx;
	const partVal = arr[endIdx];

	for (let i = begIdx; i < endIdx; i++) {
		if (arr[i] < partVal) {
			swap(arr, i, partIdx++);
		}
	}

	swap(arr, partIdx, endIdx);

	return partIdx;
};

const quicksort = (arr, begIdx = 0, endIdx = arr.length - 1) => {
	if (begIdx < endIdx) {
		const partIdx = partition(arr, begIdx, endIdx);

		quicksort(arr, begIdx, partIdx - 1);
		quicksort(arr, partIdx + 1, endIdx);
	}
};

const swap = (arr, x, y) => {
	const temp = arr[x];
	arr[x] = arr[y];
	arr[y] = temp;
};

const myArr = randomArr(true, 10);
console.log(myArr);

quicksort(myArr);

console.log(myArr);
